drivers/flash/flash_simulator.c

```diff
@@ -0,0 +1,324 @@
+	STATS_NAME(flash_sim_stats, excess_erase_cycles_unit##N)
+
+/* retrieve a unit erase cycles (EC) statistic */
+#define STATS_EC_GET(U) (*(&flash_sim_stats.erase_cycles_unit0 + (U)))
+/* increment a unit erase cycles (EC) statistic */
+#define STATS_EC_INC(U) (*(&flash_sim_stats.erase_cycles_unit0 + (U)) += 1)
+/* increment a unit excess erase cycles (EEC) statistic */
+#define STATS_EEC_INC(U)                                                       \
+	(*(&flash_sim_stats.excess_erase_cycles_unit0 + (U)) += 1)
+
+/* simulator statistcs */
+
+STATS_SECT_START(flash_sim_stats)
+STATS_SECT_ENTRY32(flash_read_calls)       /* calls to flash_read() */
+STATS_SECT_ENTRY32(flash_read_time_us)     /* time spent in flash_read() */
+STATS_SECT_ENTRY32(flash_read_api_errors)  /* flash_read() API errors */
+STATS_SECT_ENTRY32(flash_read_hw_errors)   /* hardware errors during read */
+STATS_SECT_ENTRY32(flash_write_calls)      /* calls to flash_write() */
+STATS_SECT_ENTRY32(flash_write_time_us)    /* time spent in flash_write() */
+STATS_SECT_ENTRY32(flash_write_api_errors) /* flash_write() API errors */
+STATS_SECT_ENTRY32(flash_write_hw_errors)  /* hardware errors during write */
+STATS_SECT_ENTRY32(flash_erase_calls)      /* calls to flash_erase() */
+STATS_SECT_ENTRY32(flash_erase_time_us)    /* time spent in flash_erase() */
+STATS_SECT_ENTRY32(flash_erase_api_errors) /* flash_erase() API errors */
+STATS_SECT_ENTRY32(flash_erase_hw_errors)  /* hardware errors during erase */
+STATS_SECT_ENTRY32(bytes_written)	  /* total bytes written */
+UTIL_EVAL(UTIL_REPEAT(CONFIG_FLASH_SIMULATOR_FLASH_SIZE,
+		      STATS_SECT_EC)) /* erase cycle count for unit */
+UTIL_EVAL(UTIL_REPEAT(CONFIG_FLASH_SIMULATOR_FLASH_SIZE,
+		      STATS_SECT_EEC)) /* excess erase cycle count for unit */
+STATS_SECT_END;
+
+STATS_SECT_DECL(flash_sim_stats) flash_sim_stats;
+STATS_NAME_START(flash_sim_stats)
+STATS_NAME(flash_sim_stats, flash_read_calls)
+STATS_NAME(flash_sim_stats, flash_read_time_us)
+STATS_NAME(flash_sim_stats, flash_read_api_errors)
+STATS_NAME(flash_sim_stats, flash_read_hw_errors)
+STATS_NAME(flash_sim_stats, flash_write_calls)
+STATS_NAME(flash_sim_stats, flash_write_time_us)
+STATS_NAME(flash_sim_stats, flash_write_api_errors)
+STATS_NAME(flash_sim_stats, flash_write_hw_errors)
+STATS_NAME(flash_sim_stats, flash_erase_calls)
+STATS_NAME(flash_sim_stats, flash_erase_time_us)
+STATS_NAME(flash_sim_stats, flash_erase_api_errors)
+STATS_NAME(flash_sim_stats, flash_erase_hw_errors)
+STATS_NAME(flash_sim_stats, bytes_written)
+UTIL_EVAL(UTIL_REPEAT(CONFIG_FLASH_SIMULATOR_FLASH_SIZE, STATS_NAME_EC))
+UTIL_EVAL(UTIL_REPEAT(CONFIG_FLASH_SIMULATOR_FLASH_SIZE, STATS_NAME_EEC))
+STATS_NAME_END(flash_sim_stats);
+
+static u8_t mock_flash[FLASH_SIZE];
+static bool write_protection;
+static u64_t _rand;
+
+#ifdef CONFIG_FLASH_SIMULATOR_SIMULATE_FAILURES
+static int toss(int failure_rate_pm)
+{
+	_rand = 1103515245 * _rand + 12345;
```

Comment:
The implementation of sys_rand32_get() varies across SoC/boards.
On some boards, specifically on qemu_cortex_m3, it simply wasn't random enough since it basically returned the number of cycles (which only increases) especially in a short time frame.
This is a very simple rand() implementation from glibc that I seed using sys_rand32_get() and seems to perform decently across boards and has the advantage of not requiring a specific libc implementation.

Reply:
Why not use rand() form newlibc? Haw it similar or worst quality?

Reply:
Mostly because we would force the application to use newlibc.
Although it might be perfectly acceptable when testing, it might be undesirable if a user wants to use this driver during development or debugging, since it would force them to use a different libc implementation.

I believe an alternative to this solution would be to tweak the sys_rand32_get() implementation for qemu_cortex_m3 (/subsys/random/rand32_timer.c). Using a multiplication instead of addition there seems to work better.